src/common/hist_util.cc

```diff
@@ -525,33 +586,93 @@ FindGroups(const std::vector<unsigned>& feature_list,
+        if (cnt <= rest_max_cnt) {
+          need_new_group = false;
+          groups[gid].push_back(fid);
+          group_conflict_cnt[gid] += cnt;
+          group_nnz[gid] += cur_fid_nnz - cnt;
+          MarkUsed(&conflict_marks[gid], column1);
+          break;
+        }
+      }
+      // create new group if necessary
+      if (need_new_group) {
+        groups.emplace_back();
+        groups.back().push_back(fid);
+        group_conflict_cnt.push_back(0);
+        conflict_marks.emplace_back(nrow, false);
+        MarkUsed(&conflict_marks.back(), column1);
+        group_nnz.emplace_back(cur_fid_nnz);
       }
     }
-
-    // create new group if necessary
-    if (need_new_group) {
-      groups.emplace_back();
-      groups.back().push_back(fid);
-      group_conflict_cnt.push_back(0);
-      conflict_marks.emplace_back(nrow, false);
-      MarkUsed(&conflict_marks.back(), column);
-      group_nnz.emplace_back(cur_fid_nnz);
+    if (colmat.GetTypeSize() == 2) {
+      const Column<uint16_t>& column2 = colmat.GetColumn<uint16_t>(fid);
+      // examine each candidate group: is it okay to insert fid?
+      for (auto gid : search_groups) {
+        const size_t rest_max_cnt = max_conflict_cnt - group_conflict_cnt[gid];
+        const size_t cnt = GetConflictCount(conflict_marks[gid], column2, rest_max_cnt);
+        if (cnt <= rest_max_cnt) {
+          need_new_group = false;
+          groups[gid].push_back(fid);
+          group_conflict_cnt[gid] += cnt;
+          group_nnz[gid] += cur_fid_nnz - cnt;
+          MarkUsed(&conflict_marks[gid], column2);
+          break;
+        }
+      }
+      // create new group if necessary
+      if (need_new_group) {
+        groups.emplace_back();
+        groups.back().push_back(fid);
+        group_conflict_cnt.push_back(0);
+        conflict_marks.emplace_back(nrow, false);
+        MarkUsed(&conflict_marks.back(), column2);
+        group_nnz.emplace_back(cur_fid_nnz);
+      }
+    }
+    if (colmat.GetTypeSize() == 4) {
+      const Column<uint32_t>& column4 = colmat.GetColumn<uint32_t>(fid);
+      // examine each candidate group: is it okay to insert fid?
+      for (auto gid : search_groups) {
+        const size_t rest_max_cnt = max_conflict_cnt - group_conflict_cnt[gid];
+        const size_t cnt = GetConflictCount(conflict_marks[gid], column4, rest_max_cnt);
+        if (cnt <= rest_max_cnt) {
+          need_new_group = false;
+          groups[gid].push_back(fid);
+          group_conflict_cnt[gid] += cnt;
+          group_nnz[gid] += cur_fid_nnz - cnt;
+          MarkUsed(&conflict_marks[gid], column4);
+          break;
+        }
+      }
+      // create new group if necessary
+      if (need_new_group) {
+        groups.emplace_back();
+        groups.back().push_back(fid);
+        group_conflict_cnt.push_back(0);
+        conflict_marks.emplace_back(nrow, false);
+        MarkUsed(&conflict_marks.back(), column4);
+        group_nnz.emplace_back(cur_fid_nnz);
+      }
```

Comment:
Too much copy-past code. I think you can replace this by usage of template functions...

Reply:
done